net/gclient: fix parameters containing equal signs were tampered

fghwett opened the pull request:
Fixed the issue where parameters containing equal signs were tampered with when there were file parameters in the same form.

Example:
```go
gclient.New().Post(context.Background(), "https://example.com", fmt.Sprintf("source=https://exmaple.com/rand?type=girls&image=@file:filename.png"))

// gclient_request.go:233 this will cause the parameters to be divided into three sections. [source, https://exmaple.com/rand?type, girls] 
for _, item := range strings.Split(params, "&") {
	array := strings.Split(item, "=")
	...
}

//  gclient_request.go:232 this will cause the third and subsequent parameters to be lost.
var (
	fieldName  = array[0]
	fieldValue = array[1]
)
if err = writer.WriteField(fieldName, fieldValue); err != nil {
	err = gerror.Wrapf(err, `write form field failed with "%s", "%s"`, fieldName, fieldValue)
	return nil, err
}
```

fghwett committed:
Fixed the issue where parameters containing equal signs were tampered with when there were file parameters in the same form.

gqcn commented:
@fghwett Good enhancement! Would you please add associated unit testing case for this update?

fghwett committed:
Fixed an issue where file names containing = and & and equally complex parameters were tampered with in the form request parameters.

fghwett commented:
When I was writing the test case, I found a problem with more than just containing equal signs. After looking at the code, I decided it was the '@file:' parameter and noUrlEncode, so I made some adjustments.
Now, as long as the @file: parameter is included, it will be encoded and decoded in the corresponding processing place. This is the processing scheme I have come up with at present, I don't know whether it can be accepted.

gqcn commented:
> Fixed the issue where parameters containing equal signs were tampered with when there were file parameters in the same form.
> 
> Example:
> 
> ```go
> gclient.New().Post(context.Background(), "https://example.com", fmt.Sprintf("source=https://exmaple.com/rand?type=girls&image=@file:filename.png"))
> 
> // gclient_request.go:233 this will cause the parameters to be divided into three sections. [source, https://exmaple.com/rand?type, girls] 
> for _, item := range strings.Split(params, "&") {
> 	array := strings.Split(item, "=")
> 	...
> }
> 
> //  gclient_request.go:232 this will cause the third and subsequent parameters to be lost.
> var (
> 	fieldName  = array[0]
> 	fieldValue = array[1]
> )
> if err = writer.WriteField(fieldName, fieldValue); err != nil {
> 	err = gerror.Wrapf(err, `write form field failed with "%s", "%s"`, fieldName, fieldValue)
> 	return nil, err
> }
> ```

@fghwett or you can urlencode your parameters? like:
```go
g.Client().Post(
    context.Background(), 
    "https://example.com", 
    "source=https%3A%2F%2Fexmaple.com%2Frand%3Ftype%3Dgirls%26image%3D%40file%3Afilename.png",
)
```

fghwett commented:
> > Fixed the issue where parameters containing equal signs were tampered with when there were file parameters in the same form.
> > Example:
> > ```go
> > gclient.New().Post(context.Background(), "https://example.com", fmt.Sprintf("source=https://exmaple.com/rand?type=girls&image=@file:filename.png"))
> > 
> > // gclient_request.go:233 this will cause the parameters to be divided into three sections. [source, https://exmaple.com/rand?type, girls] 
> > for _, item := range strings.Split(params, "&") {
> > 	array := strings.Split(item, "=")
> > 	...
> > }
> > 
> > //  gclient_request.go:232 this will cause the third and subsequent parameters to be lost.
> > var (
> > 	fieldName  = array[0]
> > 	fieldValue = array[1]
> > )
> > if err = writer.WriteField(fieldName, fieldValue); err != nil {
> > 	err = gerror.Wrapf(err, `write form field failed with "%s", "%s"`, fieldName, fieldValue)
> > 	return nil, err
> > }
> > ```
> 
> @fghwett or you can urlencode your parameters? like:
> 
> ```go
> g.Client().Post(
>     context.Background(), 
>     "https://example.com", 
>     "source=https%3A%2F%2Fexmaple.com%2Frand%3Ftype%3Dgirls%26image%3D%40file%3Afilename.png",
> )
> ```

I tried the following two ways and got wrong results.

```go
# Only the parameters other than the file are encoded, and the backend will receive the encoded URL. However, in terms of business logic, the backend does not perform URL de-encoding processing.
resp, err := gclient.New().
	SetHeader("token", "zk5tkqa8te7rj30").
	Post(context.Background(), "https://exmaple.com",
	fmt.Sprintf("source=%s&image=@file:%s", gurl.Encode("https://exmaple.com/rand?type=girls"), "filepath.jpg"),
)

# Encode all parameters, gclient will think they are uploaded in json format.
resp, err := gclient.New().
	Post(context.Background(), "https://example.com",
	gurl.Encode("source=https://exmaple.com/rand?type=girls&image=@file:filename.jpg"),
)
```

In the business logic, I need to upload a photo along with the original URL of the photo.

fghwett commented:
> @fghwett What you expect is the server side receiving the normal source parameter (`https://exmaple.com/rand?type=girls`) and the uploaded file, or the complete source parameter containing file parameter(`https://exmaple.com/rand?type=girls&image=@file:filename.png`)?

I'm the first case.

gqcn commented:
I'm still handling on this.

lingcoder commented:
感谢提交。这个问题在 master 上已经修复：                                                                                                                       
  net/gclient/gclient_request.go 现已使用 strings.SplitN(item, "=", 2)，                                                                                         
  并增加了 len(array) < 2 的越界保护，字段名与值也都会经过 gurl.Decode。
